text/0002-modules.md

```diff
@@ -0,0 +1,59 @@
+Variables within a module are not global-by-default and are not accessible to other javascript code unless explicitly
+exported.
+
+Javascript modules are also known as ECMAScript modules, ES modules, ESM, ES2015 modules, or ES6 modules.
+
+An in-browser javascript module is a javascript module that is downloaded in the browser. This is different from
+build-time modules, which are compiled down so that they know longer use `import` or `export` by the time they
+are executed in the browser. Most major frameworks (React, Vue, Angular, etc) always use build-time modules, but do not
+always use in-browser modules. At OpenMRS, we use both.
+
+In-browser javascript modules are downloaded with one network request per module. Build-time modules are compiled away such
+that there is only one network request for a whole group (bundle) of build-time modules.
+
+[Basic browser support for modules](https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export#Browser_compatibility)
+started in Chrome 61, Firefox 60, Edge 16, and Safari 10.1. There are advanced module features, such as import maps, that are not yet
+supported in any browser. Internet Explorer does not have any support for modules.
+
+A [module loader](https://www.jvandemo.com/a-10-minute-primer-to-javascript-modules-module-formats-module-loaders-and-module-bundlers/) is
+a javascript library that allows you to use modules in a browser that does not support them. This makes it possible for us to use both basic and advanced
+module features in all browsers.
+
+[SystemJS](https://github.com/systemjs/systemjs) is a minimalistic module loader that
+[polyfills](https://en.wikipedia.org/wiki/Polyfill_(programming)) modules so that you can use them in older browsers.
+When browsers fully support modules, the browser will be able to do what SystemJS is currently doing for us.
+
+## Reason for decision
+- Modules are a built-in programming language feature that allows us to easily extend the OpenMRS frontend.
+- Modules allow distribution implementors and maintainers to extend OpenMRS functionality via a browser spec, instead of something
+OpenMRS-specific. This means that the process for extending OpenMRS is better understood and more approachable for frontend developers.
+- Modules create better [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns), since dependencies
+between modules are explicit and do not rely on global variables.
+- Modules can be lazy loaded, or not loaded at all. This allows OpenMRS distributions to choose which modules they wish
+to include in their code.
+
+## Alternatives
+Instead of using both in-browser modules and build-time modules, we could use build-time modules exclusively. This would
+simplify some things, but would mean that all frontend code would have to go through a single build process, which would be
+very difficult to get working when supporting multiple frameworks and codebases.
```

Comment:
Agreed... seems like a maintenance nightmare!

Reply:
💯

Reply:
@joeldenning, could you give an example of where in-browser modules will simplify things?  Am I understanding right that with build process, your team needs a process to build using webpack and produce a deployable artifact, whereas with the runtime process your team needs a process to maintain an html file that is kind of a runtime equivalent of a webpack config that defines dependencies in script tags, and a main js file to execute?  In both cases, a team needs to write code to create a "distribution" of javascript modules.  I'm trying to understand how a "single build process, which would be very difficult to get working when supporting multiple frameworks and codebases" plays out in practice.  Thanks!

Reply:
> could you give an example of where in-browser modules will simplify things?

This is best explained by talking about what you have to do if you don't have in-browser modules. See my answer below about build process -- hopefully that will answer.

> Am I understanding right that with build process, your team needs a process to build using webpack and produce a deployable artifact,

Webpack is not required (rollup, parcel, jspm also fine), and actually a build process isn't strictly required at all. But yes, most modules would be built with webpack. Webpack is what angular-cli, vue-cli, and react all do by default. I plan on submitting an RFC PR in the next few days about "build process" to clarify exactly what is required.

>  whereas with the runtime process your team needs a process to maintain an html file that is kind of a runtime equivalent of a webpack config that defines dependencies in script tags, and a main js file to execute?

Yeah you have to have a map of all the in-browser modules and their urls. See [import maps](https://docs.google.com/presentation/d/1sv0n_15Zp9HNusdSagOXnBO7Kbh2qURMQj3a8OR8ndo/edit#slide=id.g5681fc88ca_0_146) and [SystemJS](https://github.com/systemjs/systemjs). My next RFC PR will be about Import Maps.

> I'm trying to understand how a "single build process, which would be very difficult to get working when supporting multiple frameworks and codebases" plays out in practice.

Let's break this down into 3 different options for how a large single page application works:

- Option 1: A single webpack build for everything
This would mean that the OpenMRS code and the distribution code would ultimately need to all be in the same file system at build time, and that there would be exactly one webpack config that everything has to live with. If we do not dictate a framework for all OpenMRS (which I do not want to do), a single webpack config takes angular-cli, vue-cli, and create-react-app off the table for people writing OpenMRS code. Reason for that is that each assumes sole and total control over their own webpack config. Those are **the absolute most popular ways to build write frontend code**, so taking them off the table would be a big restriction
- Option 2: Multiple webpack builds, but no in-browser modules
This would mean that create-react-app, angular-cli, and vue-cli are all back on the table. But now how do you get those separate webpack bundles to coordinate in the browser?? Global variables? Communicating through global variables is essentially doing in-browser modules except with implicit dependencies and load-order dependencies.
- Option 3: Multiple webpack builds, with in-browser module
You can use vue-cli, angular-cli, react, or anything else. Plus your code can coordinate/communicate with other modules that aren't even written in the same framework. Modules are universal to javascript.

**Video that will help explain maybe**
For an overview of what this architecture offers and what it could mean for OpenMRS, check out this video I made: https://drive.google.com/file/d/1OkdaDJf4lCTc1FwQcxsWxx-7Kk9WaYpW/view?usp=sharing.